hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java

```diff
@@ -171,13 +187,20 @@ public Predicate toPredicate(final Root<T> root, final CriteriaQuery<?> query, f
         private final Root<T> root;
         private final CriteriaBuilder cb;
         private final Class<A> enumType;
+        private final VirtualPropertyLookup virtualPropertyLookup;
+        private final StrSubstitutor substitutor;
 
         private final SimpleTypeConverter simpleTypeConverter;
 
-        private JpqQueryRSQLVisitor(final Root<T> root, final CriteriaBuilder cb, final Class<A> enumType) {
+        private JpqQueryRSQLVisitor(final Root<T> root, final CriteriaBuilder cb, final Class<A> enumType,
+                VirtualPropertyLookup virtualPropertyLookup) {
             this.root = root;
             this.cb = cb;
             this.enumType = enumType;
+            this.virtualPropertyLookup = virtualPropertyLookup;
```

Comment:
I guess the field `virtualPropertyLookup` can be removed from the `JpqQueryRSQLVisitor` due it's only used to do a null-check in line `450` to check if the value can be replaced. I think you can just use the `substitutor` for this purposes in case `virtualPropertyLookup` is null.

```
this.substitutor = (virtualPropertyLookup!=null) 
                             ? new StrSubstitutor(new StrLookupAdapter(virtualPropertyLookup), StrSubstitutor.DEFAULT_PREFIX, StrSubstitutor.DEFAULT_SUFFIX, StrSubstitutor.DEFAULT_ESCAPE);
                             : null;
```

An then later you can do the null-check based on the `substitutor`